src/Symfony/Component/HttpClient/Chunk/ErrorChunk.php

```diff
@@ -0,0 +1,104 @@
+<?php
+
+/*
+ * This file is part of the Symfony package.
+ *
+ * (c) Fabien Potencier <fabien@symfony.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Component\HttpClient\Chunk;
+
+use Symfony\Component\HttpClient\Exception\TransportException;
+use Symfony\Contracts\HttpClient\ChunkInterface;
+
+/**
+ * @author Nicolas Grekas <p@tchwork.com>
+ *
+ * @internal
+ */
+class ErrorChunk implements ChunkInterface
+{
+    protected $throw = true;
+
+    private $offset;
+    private $error;
+
+    public function __construct(int $offset, \Throwable $error = null)
+    {
+        $this->offset = $offset;
+        $this->error = null !== $error ? [$error->getMessage(), null, null, $error] : ['Reading from the response stream reached the inactivity timeout.'];
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function isTimeout(): bool
+    {
+        $this->throw = false;
+
+        if (1 < \count($this->error)) {
+            throw new TransportException(...$this->error);
+        }
+
+        return true;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function isFirst(): bool
+    {
+        $this->throw = false;
+        throw new TransportException(...$this->error);
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function isLast(): bool
+    {
+        $this->throw = false;
+        throw new TransportException(...$this->error);
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function getContent(): string
+    {
+        $this->throw = false;
+        throw new TransportException(...$this->error);
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function getOffset(): int
+    {
+        $this->offset;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function getError(): ?string
+    {
+        return $this->error[0];
+    }
+
+    public function __toString()
+    {
+        return '';
+    }
+
+    public function __destruct()
+    {
+        if ($this->throw) {
+            $this->throw = false;
+            throw new TransportException(...$this->error);
```

Comment:
Thanks for raising this @arnaud-lb.

This exception can be thrown from many places: public methods of the chunk (with the notable exception of `__toString` of course), but it is also thrown from the response itself when you call e.g. `getContent`. That means that if you actually use the chunk and/or the response, you'll get the error before the destructor is called.

Now, if you don't do anything with the chunk and if you don't call `getContent` (e.g. because you write some code that only waits for the response headers without checking them and neither the content - like a remote "ping"), then I still want the error to be reported so that lightly written codes are still safe. That's what this destructor ensures.

For reference (you may know it but other readers may no), about the technical behavior of destructors, its behavior is quite predictable:
- when no circular loops are involved, the destructor is called immediately when the object is not referenced anymore,
- when a circular loop is involved, this is postponed to the next run of the GC, which can be in some unrelated code file/line yes
- when a nasty fatal error occurs in the shutdown sequence, destructors are sometimes not called. Unless this happens (quite rare), destructors are always called in one of the two situations before.

So, here, the situation aligns to the requirements:
- response and chunk objects don't embed any circular loops - they will be destructed immediately when unreferenced
- destructors are "only" a safety measure to report back some error to the consumer in case one didn't directly use either the chunk or the response.

Destructors add to the robustness of the component, but they're not central to error reporting.
So all good to me :) Sold on it too?

Reply:
I agree on where and how destructors are called. My point is that these conditions will happen, and/or will be worse than you think for the usability of the component.

> That means that if you actually use the chunk and/or the response, you'll get the error before the destructor is called.

Unless you used `__toString()`, which is what the examples are using:

```
$output = fopen('output.file', 'w');

foreach ($client->stream($response) as $chunk) {
    fwrite($output, $chunk);
}
```

In this example taken from the PR description, we fully rely on ErrorChunk's destructor for throwing exceptions and saving us. However, imagine that the code continues like this:

```
do_something_after_successfully_streaming($output);

return;
```

The exception will be thrown after return, but the happy path has already been taken.

If I understand the intent correctly, the purpose of the `ErrorChunk` is to let the user access the error string, or to throw an exception if the user didn't care about checking for errors. In other words, it's trying to throw an exception unless the user explicitly checks for errors.

So, why not throw an exception earlier ? This would be safe 100% of the time, and the exception trace would be more useful (in this example, it would mention the call site of `stream()`, which is really want we want).

> For reference (you may know it but other readers may not), about the technical behavior of destructors, it is quite predictable:
> [...]
> So, here, the situation aligns to the requirements:
>  - response and chunk objects don't embed any circular loops

Response and chunk themselves don't need to be directly involved in a circular loop. They only need to be referenced by an object involved in a circular loop.

This can happen easily by accident in non trivial code.

> destructors are "only" a safety measure to report back some error to the consumer 

Then why not throw an exception earlier ?

My point is that the current design will have non obvious behaviours that could make the client very difficult to use properly, especially when wrapper in some layers.

For example, imagine a TraceableHttpClient that keeps track of the last read chunk. We will get the exception much later, when the client itself is destroyed.

Even without considering destructors: chunks might be consumed quite remotely from the `stream()` call site, and an exception calls in `getContent()` might no trace back to any useful location.

Reply:
The intended usage seems to be this:

```
foreach ($client->stream($response) as $chunk) {
    $error = $chunk->getError();
    if ($error !== null) {
        // do something with error
    }
    fwrite($output, $chunk);
}
```

And all exception traps in `getContent()`, `__destruct`, etc are there to save the user who forgot to check for `$chunk->getError()`.

I would like to suggest to throw exceptions directly from stream(), such that the intended usage would look more like this:

```
try {
    foreach ($client->stream($response) as $chunk) {
        fwrite($output, $chunk);
    }
} catch (SomeException $e) {
    // do something with error
}
```

Reply:
you're correct about the expected usage - what the destructor provides is safety for those who don't add the try/catch.

> I would like to suggest to throw exceptions directly from stream()

That would break streaming concurrent responses: when only one response stream breaks, we should be able to continue streaming the other responses. Having the `foreach` itself throw would make this much harder, if possible at all.

Reply:
> imagine a TraceableHttpClient that keeps track of the last read chunk. We will get the exception much later, when the client itself is destroyed

the exception has a previous one attached (`$e->getPrevious()`) that will give the stack trace at the actual error time, so that this delay shouldn't be an issue.

Reply:
This delay will be an issue, because the exception will happen when it doesn't matter anymore:

```
    foreach ($client->stream($response) as $chunk) {
        fwrite($output, $chunk);
    }

    do_something_after_successfully_streaming($output)
```

If the exception happens after this, it's already too late for it to be any useful.

Reply:
Hum, the issue is `__toString` indeed. We should replace it then, `->toString()`.

Reply:
> If the exception happens after this, it's already too late for it to be any useful.

Thanks for spotting this. This is now fixed by [throwing in the stream() method](https://github.com/symfony/symfony/pull/30413/commits/95de266aad1f0de3ecbf809c99591b2c7c1d6cf2).
To allow people to continue iterating over non-errored streams, the exception is thrown after the `yield`, and only if it didn't throw already. The destructor is still used as a last resort measure to ensure the error is always reported back.

Reply:
`__toString()` is now removed from chunks, `getContent()` is the way to go.